physical/cache.go

```diff
@@ -34,16 +43,58 @@ func NewCache(b Backend, size int, logger log.Logger) *Cache {
 	c := &Cache{
 		backend: b,
 		lru:     cache,
+		locks:   make(map[string]*sync.RWMutex, 256),
+		logger:  logger,
 	}
+	if err := locksutil.CreateLocks(c.locks, 256); err != nil {
+		logger.Error("physical/cache: error creating locks", "error", err)
+		return nil
+	}
+
+	if txnl, ok := c.backend.(Transactional); ok {
+		c.transactional = txnl
+	}
+
 	return c
 }
 
+func (c *Cache) lockHashForKey(key string) string {
+	hf := sha1.New()
+	hf.Write([]byte(key))
+	return strings.ToLower(hex.EncodeToString(hf.Sum(nil))[:2])
+}
+
+func (c *Cache) lockForKey(key string) *sync.RWMutex {
+	return c.locks[c.lockHashForKey(key)]
+}
+
 // Purge is used to clear the cache
 func (c *Cache) Purge() {
+	// Lock the world
+	var lockHashes []string
+	for hash := range c.locks {
+		lockHashes = append(lockHashes, hash)
+	}
+
+	// Sort and deduplicate. This ensures we don't try to grab the same lock
+	// twice, and enforcing a sort means we'll not have multiple goroutines
+	// deadlock by acquiring in different orders.
+	lockHashes = strutil.RemoveDuplicates(lockHashes)
```

Comment:
No, that's a bit of copypasta, but I decided to leave it in. There *can* be duplicates below where this is used again. We do want it sorted here, though!